File: app/src/main/java/com/lxy/molweightcalculator/parsing/Parser.java
package com.lxy.molweightcalculator.parsing;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.lxy.molweightcalculator.BuildConfig;
import com.lxy.molweightcalculator.contract.Contract;
import com.lxy.molweightcalculator.ui.StatisticsItem;

import java.util.ArrayList;

import timber.log.Timber;

// Handwritten recursive descent/LL(1) parser
// Grammar:
// Non-terminals:
//     Formula = Compound+
//     Compound = Atom Quantity?
//                | "(" Compound ")" Quantity?
//                | "[" Compound "]" Quantity?
//                | "{" Compound "}" Quantity?
// Terminals:
//     Atom: ("A" .. "Z")("a" .. "z")?
//     Quantity: ("0" .. "9")+
public class Parser {
    @NonNull
    private static final Parser PARSER = new Parser();
    @NonNull
    private final ParseStack stack = new ParseStack();
    private CharSequence formula;
    private int length;
    private int i;

    @NonNull
    public static ParseResult parse(@NonNull CharSequence formula) {
        var parser = PARSER;
        parser.init(formula);
        return parser.parse();
    }

    private void init(@NonNull CharSequence formula) {
        this.formula = Contract.requireNonNull(formula);
        this.length = formula.length();
        this.stack.clear();
        this.i = 0;
    }

    private void tryPurgeMemory() {
        if (MemoryUsage.shouldPurge()) {
            stack.purgeMemory();
            MemoryUsage.reset();
        }
    }

    private long parseQuantity() {
        final int RADIX = 10;
        final long LIMIT = -Long.MAX_VALUE;
        final long MULTIPLY_LIMIT = LIMIT / RADIX;

        var count = 0;
        var value = 0L;
        while (i < length) {
            var ch = formula.charAt(i);
            if (ch < '0' || ch > '9') {
                break;
            } else {
                count++;
                i++;
                if (value < MULTIPLY_LIMIT) {
                    return -1;
                }
                var digit = ch - '0';
                value *= RADIX;
                if (value < LIMIT + digit) {
                    return -1;
                }
                value -= digit;
            }
        }
        return count == 0 ? 1 : -value;
    }

    private char parseElementId() {
        var firstChar = formula.charAt(i);
        if (firstChar < 'A' || firstChar > 'Z') {
            return Character.MAX_VALUE;
        }
        if (++i == length) {
            // EOF
            return Element.parseElementId(firstChar);
        }
        var secondChar = formula.charAt(i);
        if (secondChar < 'a' || secondChar > 'z') {
            return Element.parseElementId(firstChar);
        }
        i++;
        return Element.parseElementId(firstChar, secondChar);
    }

    private void handleLeftBracket(@Bracket int bracket) {
        stack.push(bracket, i++);
    }

    @Nullable
    private ParseResult handleRightBracket(@Bracket int leftBracket, @Bracket int rightBracket) {
        var top1 = stack.pop();
        var top2 = stack.peekNoThrow();
        if (top2 == null || top1.getBracket() != leftBracket) {
            return new ParseResult(i, rightBracket, ErrorCode.MISMATCHED_BRACKET);
        }
        if (top1.size() == 0) {
            return new ParseResult(i - 1, i, ErrorCode.NO_ELEMENT);
        }
        i++;
        var start = i;
        var count = parseQuantity();
        if (count < 0) {
            return new ParseResult(start, i, ErrorCode.ELEMENT_COUNT_TOO_LARGE);
        }
        if (!top2.merge(top1, count)) {
            return ParseResult.ELEMENT_COUNT_OVERFLOW;
        }
        var weight = top2.getWeight() + top1.getWeight() * count;
        if (!Double.isFinite(weight)) {
            return ParseResult.ELEMENT_COUNT_OVERFLOW;
        }
        top2.setWeight(weight);
        return null;
    }

    @NonNull
    private ParseResult parse() {
        if (length == 0) {
            if (BuildConfig.DEBUG) {
                Timber.d("Empty formula");
            }
            return ParseResult.EMPTY_FORMULA;
        }
        stack.push(ParseState.DEFAULT_BRACKET, ParseState.DEFAULT_START);
        while (i < length) {
            // Inlined parseCompound
            var start = i;
            var elementId = parseElementId();
            if ((short) elementId >= 0) {
                var state = stack.peek();
                var quantity = parseQuantity();
                if (quantity < 0) {
                    return new ParseResult(start, i, ErrorCode.ELEMENT_COUNT_TOO_LARGE);
                }
                if (!state.addValueOrPut(elementId, quantity)) {
                    return ParseResult.ELEMENT_COUNT_OVERFLOW;
                }
                var elementWeight = Element.getWeightFromId(elementId);
                if (Double.isNaN(elementWeight)) {
                    if (BuildConfig.DEBUG) {
                        Timber.d("Invalid element: %s",
                                Element.getElementNameFromId(elementId));
                    }
                    return new ParseResult(start, i, ErrorCode.INVALID_ELEMENT);
                }
                var weight = state.getWeight() + elementWeight * quantity;
                if (!Double.isFinite(weight)) {
                    return ParseResult.ELEMENT_COUNT_OVERFLOW;
                }
                state.setWeight(weight);
            } else {
                var ch = formula.charAt(i);
                switch (ch) {
                    case '(' -> handleLeftBracket(Bracket.LEFT_BRACKET);
                    case '[' -> handleLeftBracket(Bracket.LEFT_SQUARE_BRACKET);
                    case '{' -> handleLeftBracket(Bracket.LEFT_CURLY_BRACKET);
                    case ')' -> {
                        var parseResult = handleRightBracket(Bracket.LEFT_BRACKET, Bracket.RIGHT_BRACKET);
                        if (parseResult != null) {
                            return parseResult;
                        }
                    }
                    case ']' -> {
                        var parseResult = handleRightBracket(Bracket.LEFT_SQUARE_BRACKET, Bracket.RIGHT_SQUARE_BRACKET);
                        if (parseResult != null) {
                            return parseResult;
                        }
                    }
                    case '}' -> {
                        var parseResult = handleRightBracket(Bracket.LEFT_CURLY_BRACKET, Bracket.RIGHT_CURLY_BRACKET);
                        if (parseResult != null) {
                            return parseResult;
                        }
                    }
                    default -> {
                        return new ParseResult(i, i + 1, ErrorCode.INVALID_TOKEN);
                    }
                }
            }
        }
        var state = stack.pop();
        if (!stack.isEmpty()) {
            return new ParseResult(state.getStart(), state.getBracket(),
                    ErrorCode.MISMATCHED_BRACKET);
        }
        tryPurgeMemory();
        var weight = state.getWeight();
        var list = new ArrayList<StatisticsItem>(state.size());
        var size = state.size();
        for (var i = 0; i < size; i++) {
            list.add(i, new StatisticsItem(state.keyAt(i), state.valueAt(i)));
        }
        return new ParseResult(list, weight);
    }
}

File: app/src/main/java/com/lxy/molweightcalculator/parsing/Parser.kt
package com.lxy.molweightcalculator.parsing

import com.lxy.molweightcalculator.BuildConfig
import com.lxy.molweightcalculator.contract.Contract
import com.lxy.molweightcalculator.util.Utility
import timber.log.Timber

// Handwritten recursive descent/LL(1) parser
// Grammar:
// Non-terminals:
//     Formula = Compound+
//     Compound = Atom Quantity?
//                | "(" Compound ")" Quantity?
//                | "[" Compound "]" Quantity?
//                | "{" Compound "}" Quantity?
// Terminals:
//     Atom: ("A" .. "Z")("a" .. "z")?
//     Quantity: ("0" .. "9")+
class Parser(private val formula: CharSequence) {
    // Parse stack
    private var parseStack = arrayOfNulls<ParseState?>(Utility.INITIAL_CAPACITY)
    private var stackTop = -1

    // Parse information
    private var length = formula.length
    private var index = 0

    companion object {
        private const val RADIX = 10L
        private const val LIMIT = -Long.MAX_VALUE
        private const val MULTIPLY_LIMIT = LIMIT / RADIX
    }

    private fun checkStack() {
        if (stackIsEmpty) {
            Contract.fail("Stack is empty")
        }
    }

    private val stackIsEmpty get() = stackSize() < 0

    private fun stackGetItem(index: Int): ParseState {
        return parseStack[index]!!
    }

    private fun stackPeekUnchecked(): ParseState {
        return stackGetItem(stackTop)
    }

    private fun stackPeek(): ParseState {
        checkStack()
        return stackPeekUnchecked()
    }

    private fun stackPop(): ParseState {
        checkStack()
        return stackGetItem(stackTop--)
    }

    private fun stackPeekNoThrow(): ParseState? {
        return if (stackIsEmpty) null else stackPeekUnchecked()
    }

    private fun stackSize(): Int {
        return stackTop
    }

    private fun stackPush(bracket: Bracket, start: Int) {
        val s = stackSize() + 1
        if (s == parseStack.size) {
            parseStack = parseStack.copyOf(Utility.growSize(s))
        }
        val state = parseStack[s]
        if (state != null) {
            state.reset(bracket, start)
        } else {
            parseStack[s] = ParseState(bracket, start)
        }
        stackTop = s
    }

    private fun parseQuantity(): Long {
        var count = 0
        var value = 0L
        while (index < length) {
            val ch = formula[index]
            if (ch < '0' || ch > '9') {
                break
            } else {
                count++
                index++
                if (value < MULTIPLY_LIMIT) {
                    return -1
                }
                val digit = (ch.code - '0'.code).toLong()
                value *= RADIX
                if (value < LIMIT + digit) {
                    return -1
                }
                value -= digit
            }
        }
        return if (count == 0) 1 else -value
    }

    private fun parseElementId(): ElementId {
        val firstChar = formula[index]
        if (firstChar < 'A' || firstChar > 'Z') {
            return ElementId.INVALID
        }
        if (++index == length) {
            // EOF
            return ElementId.valueOf(firstChar)
        }
        val secondChar = formula[index]
        if (secondChar < 'a' || secondChar > 'z') {
            return ElementId.valueOf(firstChar)
        }
        index++
        return ElementId.valueOf(firstChar, secondChar)
    }

    private fun handleLeftBracket(bracket: Bracket) {
        stackPush(bracket, index++)
    }

    private fun handleRightBracket(
        leftBracket: Bracket,
        rightBracket: Bracket,
        parseResult: ParseResult
    ): Boolean {
        val top1 = stackPop()
        val top2 = stackPeekNoThrow()
        if (top2 == null || top1.bracket != leftBracket) {
            parseResult.init(index, rightBracket.ordinal, ErrorCode.MismatchedBracket)
            return false
        }
        if (top1.size() == 0) {
            parseResult.init(index - 1, index, ErrorCode.NoElement)
            return false
        }
        index++
        val start = index
        val count = parseQuantity()
        if (count < 0) {
            parseResult.init(start, index, ErrorCode.ElementCountTooLarge)
            return false
        }
        if (!top2.merge(top1, count)) {
            parseResult.init(ErrorCode.ElementCountOverflow)
            return false
        }
        val weight = top2.weight + top1.weight * count
        if (!weight.isFinite()) {
            parseResult.init(ErrorCode.WeightOverflow)
            return false
        }
        top2.weight = weight
        return true
    }

    fun parse(parseResult: ParseResult) {
        if (length == 0) {
            parseResult.init(ErrorCode.EmptyFormula)
            return
        }
        if (!ParseResult.canParse(length)) {
            parseResult.init(ErrorCode.FormulaTooLong)
            return
        }
        stackPush(Bracket.Invalid, -1)
        while (index < length) {
            // Inlined parseCompound
            val start = index
            val elementId = parseElementId()
            if (elementId.isValid) {
                val state = stackPeek()
                val quantity = parseQuantity()
                if (quantity < 0) {
                    parseResult.init(start, index, ErrorCode.ElementCountTooLarge)
                    return
                }
                if (!state.addValueOrPut(elementId.value, quantity)) {
                    parseResult.init(ErrorCode.ElementCountOverflow)
                    return
                }
                val elementWeight = elementId.weight
                if (elementWeight.isNaN()) {
                    if (BuildConfig.DEBUG) {
                        Timber.d("Invalid element: %s", elementId.elementName)
                    }
                    parseResult.init(start, index, ErrorCode.InvalidElement)
                    return
                }
                val weight = state.weight + elementWeight * quantity
                if (!weight.isFinite()) {
                    parseResult.init(ErrorCode.WeightOverflow)
                    return
                }
                state.weight = weight
            } else {
                val ch = formula[index]
                when (ch) {
                    '(' -> handleLeftBracket(Bracket.LeftBracket)
                    '[' -> handleLeftBracket(Bracket.LeftSquareBracket)
                    '{' -> handleLeftBracket(Bracket.LeftCurlyBracket)
                    ')' -> {
                        if (!handleRightBracket(
                                Bracket.LeftBracket,
                                Bracket.RightBracket,
                                parseResult
                            )
                        ) {
                            return
                        }
                    }

                    ']' -> {
                        if (!handleRightBracket(
                                Bracket.LeftSquareBracket,
                                Bracket.RightSquareBracket,
                                parseResult
                            )
                        ) {
                            return
                        }
                    }

                    '}' -> {
                        if (!handleRightBracket(
                                Bracket.LeftCurlyBracket,
                                Bracket.RightCurlyBracket,
                                parseResult
                            )
                        ) {
                            return
                        }
                    }

                    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' -> {
                        parseResult.init(index, index + 1, ErrorCode.NoElement)
                        return
                    }

                    else -> {
                        parseResult.init(index, index + 1, ErrorCode.InvalidToken)
                        return
                    }
                }
            }
        }
        val state = stackPop()
        if (!stackIsEmpty) {
            parseResult.init(
                state.start, state.bracket.ordinal,
                ErrorCode.MismatchedBracket
            )
            return
        }
        parseResult.init(state)
    }
}